src/parser.rs

```diff
@@ -330,75 +332,123 @@ pub fn video_data_header(input: &[u8]) -> IResult<&[u8], VideoDataHeader> {
+}
+
+#[derive(Debug, PartialEq)]
 pub enum ScriptDataValue<'a> {
-  Number,
-  Boolean,
+  Number(f64),
+  Boolean(bool),
   String(&'a str),
-  Object,
-  MovieClip,
+  Object(Vec<ScriptDataObject<'a>>),
+  MovieClip(&'a str),
   Null,
-  UNdefined,
-  Reference,
-  ECMAArray,
-  StrictArray,
-  Date,
+  Undefined,
+  Reference(u16),
+  ECMAArray(Vec<ScriptDataObject<'a>>),
+  StrictArray(Vec<ScriptDataObject<'a>>),
+  Date(ScriptDataDate),
   LongString(&'a str),
 }
 
-/*
 named!(pub script_data_object<ScriptDataObject>,
-  chain!(
-    name: script_data_string ~
-    data: script_data_value  ,
-    || {
-      ScriptDataObject {
+  do_parse!(
+    name: script_data_string >>
+    data: get_data_value     >>
+    (ScriptDataObject {
         name: name,
-        data: data
-      }
-    }
+        data: data,
+    })
   )
 );
 
-pub fn script_data_object_end(input:&[u8]) -> IResult<&[u8],()> {
-  match be_u24(input) {
-    IResult::Done(i,o) => if o == 9 {
-      IResult::Done(i,())
-    } else {
-      IResult::Error(Err::Code(ErrorKind::Tag))
-    },
-    e => e
-  }
+
+pub fn script_data_object_end(input:&[u8]) -> IResult<&[u8],&[u8]> {
+  static tag: &'static [u8] = &[0u8, 0u8, 9u8];
+  tag!(tag, input)
 }
 
 named!(pub script_data_string<&str>, map_res!(length_bytes!(be_u16), from_utf8));
 named!(pub script_data_long_string<&str>, map_res!(length_bytes!(be_u32), from_utf8));
+named!(pub script_data_date<ScriptDataDate>,
+  do_parse!(
+    date_time: be_f64               >>
+    local_date_time_offset: be_i16  >>
+    (ScriptDataDate {
+        date_time: date_time,
+        local_date_time_offset: local_date_time_offset,
+    })
+  )
+);
+named!(pub script_data_objects<Vec<ScriptDataObject> >,
+  terminated!(many0!(do_parse!(
+    name: script_data_string >>
+    value: get_data_value    >>
+    (ScriptDataObject {
+        name: name,
+        data: value,
+    })
+    )), script_data_object_end)
+);
+named!(pub script_data_ECMA_array<Vec<ScriptDataObject> >,
+  do_parse!(
+    be_u32                 >>
+    v: script_data_objects >>
+    (v)
+  )
+);
+pub fn script_data_strict_array(input: &[u8]) -> IResult<&[u8], Vec<ScriptDataObject>> {
+  match be_u32(input) {
+    IResult::Done(i, o) => many_m_n!(i, 1, o as usize, script_data_object),
+    IResult::Incomplete(i) => IResult::Incomplete(i),
+    IResult::Error(i) => IResult::Error(i),
+  }
+}
 
-named!(pub script_data_value<ScriptDataValue>, );
+named!(pub get_data_value<ScriptDataValue>,
+  switch!(be_u8,
+      0  => map!(be_f64, |n| ScriptDataValue::Number(n))
+    | 1  => map!(be_u8, |n| ScriptDataValue::Boolean(n != 0))
+    | 2  => map!(script_data_string, |n| ScriptDataValue::String(n))
+    | 3  => map!(script_data_objects, |n| ScriptDataValue::Object(n))
+    | 4  => map!(script_data_string, |n| ScriptDataValue::MovieClip(n))
+    | 5  => value!(ScriptDataValue::Null) // to remove
+    | 6  => value!(ScriptDataValue::Undefined) // to remove
```

Comment:
Undefined and Null are both not in the spec. Why were they here? Should they just be removed as the comment suggests?